### PR TITLE
[6024] changed the state to work ui according to system

### PR DESCRIPTION
**Pull request by akshay-bharadva**
Close: #6024 

### Summary
🖱️🎨♻️

Added and improved the resource name editing feature in the `ResourceHome` component. Refactored the `ResourceNameField` component to make it more reusable and maintainable. Updated the styling and layout of the resource name field and its icons.

> _Oh, we're the coders of the sea, and we work on the UI_
> _We make the resource name field easy to modify_
> _We style it and refactor it with a clever trick_
> _We use the hover state to show the edit icon_

### Walkthrough
*  Import `useState` hook from React to use state variables in `ResourceHome` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-1b796664945ea9810e1092727a9e0320e9697e68cfaed66a2a68d5a1f773ec15L4-R4))
* Add `hovered` state variable and `handleMouseEnter` and `handleMouseLeave` functions to track mouse hover over resource name field in `ResourceHome` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-1b796664945ea9810e1092727a9e0320e9697e68cfaed66a2a68d5a1f773ec15R37-R46))
* Add `onMouseEnter` and `onMouseLeave` props to `div` element that wraps `ResourceNameField` component to trigger hover functions in `ResourceHome` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-1b796664945ea9810e1092727a9e0320e9697e68cfaed66a2a68d5a1f773ec15L52-R74))
* Pass `hovered` and `hoverHandler` props to `ResourceNameField` component to pass `hovered` state and `setHovered` function from `ResourceHome` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-1b796664945ea9810e1092727a9e0320e9697e68cfaed66a2a68d5a1f773ec15L52-R74))
* Add style attribute to `div` element that wraps `CircleBadge` component for debugging purposes in `ResourceHome` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-1b796664945ea9810e1092727a9e0320e9697e68cfaed66a2a68d5a1f773ec15L52-R74))
* Add CSS class to style container of icons next to resource name field in `ResourceNameField.scss` ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-3c440f7656707852f3468f1a648bac07858594e8c748398285ba8285c01b6df3R2-R4))
* Modify CSS classes for `form` and `input` elements to change layout and alignment of resource name field when editing in `ResourceNameField.scss` ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-3c440f7656707852f3468f1a648bac07858594e8c748398285ba8285c01b6df3L7-R18))
* Modify CSS class for `edit` element to change alignment of resource name field when not editing in `ResourceNameField.scss` ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-3c440f7656707852f3468f1a648bac07858594e8c748398285ba8285c01b6df3L35-R42))
* Add `hovered` and `hoverHandler` props to `Props` type of `ResourceNameField` component to receive `hovered` state and `setHovered` function from parent component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfR17-R18))
* Destructure `hovered` and `hoverHandler` props from `Props` object in `ResourceNameField` component and remove `hovered` state variable and `setHovered` function as they are no longer needed in this component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfL36-R45))
* Call `hoverHandler` function with `false` instead of `setHovered` when input field is blurred to update parent component's state in `ResourceNameField` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfL54-R60))
* Remove `handleMouseEnter` and `handleMouseLeave` functions from `ResourceNameField` component as they are no longer needed in this component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfL80-L86))
* Change JSX structure of `ResourceNameField` component to separate icons from form value in different `div` elements and use `hovered` prop to conditionally render edit icon only when mouse is hovering and field is not editing ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfL98-R148))
* Remove `onMouseEnter` and `onMouseLeave` props from `edit` element as they are handled by parent component in `ResourceNameField` component ([link](https://github.com/amplication/amplication/pull/6192/files?diff=unified&w=0#diff-d14ce2186a15b8b46ecf5b88e7da383cae71947f5b352c44deddb9fd47f5f7bfL98-R148))

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

**Comment by GreenMachine01** [comment]
@abrl91 please assist.

**Comment by abrl91** [comment]
@akshay-bharadva 
I think that this is the regex pattern: `/[^\w\s-]/g`

Check if a string contains no disallowed special characters:

```
function isValid(input) {
    const pattern = /[^\w\s-]/g;
    return !pattern.test(input);
}
```

so for "n%me" `isValid` will return false

**Comment by akshay-bharadva** [comment]
@abrl91 

I got the idea about validation through regx but have a doubt regarding how I show invalid character in error color?

**Comment by souravjain540** [comment]
@barshimi can you please reply

File: packages/amplication-client/src/Resource/ResourceHome.tsx [hunk 1]
```diff
@@ -49,12 +59,19 @@ const ResourceHome = ({ match, innerRoutes }: Props) => {
               backgroundColor:
                 resourceThemeMap[currentResource?.resourceType].color,
             }}
+            onMouseEnter={handleMouseEnter}
+            onMouseLeave={handleMouseLeave}
           >
             <ResourceNameField
+              hovered={hovered}
```
Comment (barshimi) [inline packages/amplication-client/src/Resource/ResourceHome.tsx:66]: why do you pass the set and the value to new component on props and not create them inside the component who uses them

File: packages/amplication-client/src/Resource/ResourceHome.tsx [hunk 2]
```diff
@@ -49,12 +59,19 @@ const ResourceHome = ({ match, innerRoutes }: Props) => {
...
+              hoverHandler={setHovered}
               currentResource={currentResource}
               resourceId={resourceId}
             />
-            <div className="circle-badge-container">
+            <div
+              className="circle-badge-container"
+              style={{ border: "1px solid red" }}
```
Comment (barshimi) [inline packages/amplication-client/src/Resource/ResourceHome.tsx:73]: why inline style and not part of the css

File: packages/amplication-client/src/Resource/ResourceNameField.tsx [hunk 3]
```diff
@@ -14,6 +14,8 @@ import "./ResourceNameField.scss";
 type Props = {
   currentResource: models.Resource;
   resourceId: string;
+  hovered: boolean;
```
Comment (barshimi) [inline packages/amplication-client/src/Resource/ResourceNameField.tsx:17]: this should be part of the component

File: packages/amplication-client/src/Resource/ResourceHome.tsx (hunk 2 above)
Comment (akshay-bharadva) [reply packages/amplication-client/src/Resource/ResourceHome.tsx]: Yes, I've removed this style, it was just for debugging purpose

File: packages/amplication-client/src/Resource/ResourceNameField.tsx (hunk 3 above)
Comment (akshay-bharadva) [reply packages/amplication-client/src/Resource/ResourceNameField.tsx:17]: The issue was that I didn't get the flag while hovering over the banner, but with prop passing I'm able to show the edit icon in the project name while hovering over the banner.

File: packages/amplication-client/src/Resource/ResourceHome.tsx (hunk 1 above)
Comment (akshay-bharadva) [reply packages/amplication-client/src/Resource/ResourceHome.tsx:66]: To get the hovered state in the Formik input for edit icon.

**Pull request closed without merging**